### DISCUSSION TITLE
[Request] Keyboard shortcut parity with notebook classic

### DISCUSSION THREAD

**Discussion by a3lem** [jupyter/notebook#6591] [Ideas]
Excited to learn that future versions of notebook will be based off the same core as JupyterLab. I'm creating this post as a user who has tried to embrace JupyterLab but, despite all its great extra features, has always returned to notebook classic. 

The reason: JupyterLab's keyboard shortcuts different enough that they frustrate many years of muscle memory built up in notebook classic. Sure, I can configure them to what I'm used to, but that means additional work each time I want to set up a (sometimes one-off) Jupyter environment somewhere. 

Having just tried the Notebook 7 alpha, I can already tell the shortcuts are more like those in JupyterLab. I'm curious, does anyone know why JupyterLab didn't just keep the same shortcuts as before? It's always striked me as such a weird UX decision.

As for the question in my title, I'd like to float the idea of reversing that move in Notebook 7. Sooner or later, notebook classic users will be upgrading to notebook 7 without realizing the important changes beneath the surface. Doubtless many of them will be less than excited to learn they need to learn new shortcuts to get their work done faster.

Or is it just me?

**Comment by jtpio** [comment]
Many thanks @lemontheme for testing the Notebook 7 pre-releases and opening this discussion, this is valuable feedback :+1: 

In JupyterLab and Notebook 7 keyboard shortcuts are configurable by the users. So you should already be able to tweak them via the settings editor.

That said there is indeed a case for keeping the same shortcuts as in the classic notebook by default. Maybe Notebook 7 should do that by providing a list of overrides in its settings. Especially if that can help make the transition from 6 to 7 smoother.

For reference there is already an shortcut override in Notebook 7 that is not in JupyterLab:

https://github.com/jupyter/notebook/blob/0dc766e88284ea14dc7c5efb4f4ce595559473f1/packages/documentsearch-extension/schema/notebookShellWidgetListener.json#L4-L11

Is there a subset of shortcuts you are particularly interested in keeping? Or does this apply to *all* the shortcuts?

**Reply by a3lem** [reply]
*All* the shortcuts, I would argue. At least, that's what makes most sense to me, from the perspective of smoothening the transition for NB classic users.

I should have included a few example shortcuts in my original post. To get a better overview, I took a screenshot of the NB classic help modal (the one that pops up when you press <kbd>h</kbd>, which – incidentally – doesn't work in NB7 either). The shortcuts that are different in NB7 are highlighted in red.

<img width="541" alt="command mode red means different" src="https://user-images.githubusercontent.com/15968876/196710364-060f9ccf-7768-4a98-914a-c2852dd0ccf6.png">
<img width="548" alt="edit mode  red means different" src="https://user-images.githubusercontent.com/15968876/196710379-237cb1c1-fa31-4a13-8844-b851ec288816.png">

Couple of additional remarks, for completeness.

> In JupyterLab and Notebook 7 keyboard shortcuts are configurable by the users

That doesn't appear to be the case in NB7. It looks like the only way to do this right now is to switch the interface to JupyterLab.

> Maybe Notebook 7 should do that by providing a list of overrides in its settings.

I like this idea, although it could get confusing when switching between the Notebook and the JupyterLab interfaces. I noticed that shortcuts that I'd set in JupyterLab months ago were being used in NB7. 

A different take on this might be to provide some sort of toggle that allows users to switch between 'Classic mode' and 'JupyterLab mode'.

> there is already an shortcut override in Notebook 7

Just FYI, the shortcut in question doesn't work on macOS. In command mode, it triggers the browser's text search/'Find' tool. In edit mode, it opens a cell-bound find-and-replace widget (with unresponsive text fields).

Although it illustrates something I've never understood about JupyterLab's shortcuts: the reliance on modifier keys like 'accel', 'alt', etc., with their obvious potential of conflicting with OS or browser shortcuts.

**Reply by jtpio** [reply]
> That doesn't appear to be the case in NB7. It looks like the only way to do this right now is to switch the interface to JupyterLab.

Good point, the settings editor should soon be added to Notebook 7: https://github.com/jupyter/notebook/pull/6316

>  it could get confusing when switching between the Notebook and the JupyterLab interfaces.

Agree, it would be nice to keep something coherent between lab and notebook 7.

> I noticed that shortcuts that I'd set in JupyterLab months ago were being used in NB7.

RIght. I believe it should be fixed in the latest pre-releases now (fixed in https://github.com/jupyterlab/jupyterlab/pull/11938).

> Just FYI, the shortcut in question doesn't work on macOS. In command mode, it triggers the browser's text search/'Find' tool. In edit mode, it opens a cell-bound find-and-replace widget (with unresponsive text fields).

Oh right, I meant to show this example for overriding the shortcut and *disable* it (it's still enabled in JupyterLab). Showing the browser find tool is indeed expected.

**Comment by jtpio** [comment]
@lemontheme if you feel comfortable with this, would you like to open a draft PR to start adding these shortcuts?

Otherwise, let's open an issue (or convert this discussion to an issue) so we can add it to the `7.0` milestone.

Thanks!

**Reply by a3lem** [reply]
I'd be happy to contribute in a more active sense. New to the codebase though, so I'll need some time to get set up. Should be able to make time after the weekend.

Am I correct in assuming that proposed shortcuts should be registered in a json under */packages/notebook/* on the key `jupyter.lab.shortcuts`?

**Reply by jtpio** [reply]
Right, probably this could be defined in a new schema file under https://github.com/jupyter/notebook/tree/main/packages/notebook-extension/schema and associate it with a plugin.

That way it's still configurable so folks can keep the lab shortcuts if they want to.

**Comment by jtpio** [comment]
Converting this discussion to an issue to add it to the 7.0 milestone.
